openmc/data/angle_distribution.py

```diff
@@ -0,0 +1,310 @@
+        interpolation = np.empty(len(mu_tabular), dtype=int)
+        j = 0
+
+        # Populate offsets and pairs array
+        for i, mu_i in enumerate(mu_tabular):
+            n = len(mu_i.x)
+            offsets[i] = j
+            interpolation[i] = 1 if mu_i.interpolation == 'histogram' else 2
+            pairs[0, j:j+n] = mu_i.x
+            pairs[1, j:j+n] = mu_i.p
+            pairs[2, j:j+n] = mu_i.c
+            j += n
+
+        # Create dataset for distributions
+        dset = group.create_dataset('mu', data=pairs)
+
+        # Write interpolation as attribute
+        dset.attrs['offsets'] = offsets
+        dset.attrs['interpolation'] = interpolation
+
+    @classmethod
+    def from_hdf5(cls, group):
+        """Generate angular distribution from HDF5 data
+
+        Parameters
+        ----------
+        group : h5py.Group
+            HDF5 group to read from
+
+        Returns
+        -------
+        openmc.data.AngleDistribution
+            Angular distribution
+
+        """
+        energy = group['energy'][()]
+        data = group['mu']
+        offsets = data.attrs['offsets']
+        interpolation = data.attrs['interpolation']
+
+        mu = []
+        n_energy = len(energy)
+        for i in range(n_energy):
+            # Determine length of outgoing energy distribution and number of
+            # discrete lines
+            j = offsets[i]
+            if i < n_energy - 1:
+                n = offsets[i+1] - j
+            else:
+                n = data.shape[1] - j
+
+            interp = INTERPOLATION_SCHEME[interpolation[i]]
+            mu_i = Tabular(data[0, j:j+n], data[1, j:j+n], interp)
+            mu_i.c = data[2, j:j+n]
+
+            mu.append(mu_i)
+
+        return cls(energy, mu)
+
+    @classmethod
+    def from_ace(cls, ace, location_dist, location_start):
+        """Generate an angular distribution from ACE data
+
+        Parameters
+        ----------
+        ace : openmc.data.ace.Table
+            ACE table to read from
+        location_dist : int
+            Index in the XSS array corresponding to the start of a block,
+            e.g. JXS(9).
+        location_start : int
+            Index in the XSS array corresponding to the start of an angle
+            distribution array
+
+        Returns
+        -------
+        openmc.data.AngleDistribution
+            Angular distribution
+
+        """
+        # Set starting index for angle distribution
+        idx = location_dist + location_start - 1
+
+        # Number of energies at which angular distributions are tabulated
+        n_energies = int(ace.xss[idx])
+        idx += 1
+
+        # Incoming energy grid
+        energy = ace.xss[idx:idx + n_energies]*EV_PER_MEV
+        idx += n_energies
+
+        # Read locations for angular distributions
+        lc = ace.xss[idx:idx + n_energies].astype(int)
+        idx += n_energies
+
+        mu = []
+        for i in range(n_energies):
```

Comment:
Not a fan of all the magic numbers used in this loop. Consider using variables (or providing comments explaining values)